moped-editor/src/views/projects/projectView/ProjectComponents/utils/form.js

```diff
@@ -74,12 +76,29 @@ export const useInitialValuesOnAttributesEdit = (
         (option) => option.value === initialFormValues.component.component_id
       ).label,
       data: {
-        moped_subcomponents:
-          initialFormValues.component.moped_components.moped_subcomponents,
+        // Include component subcomponents and metadata about the internal_table needed for the form
+        ...initialFormValues.component.moped_components,
       },
     });
   }, [componentOptions, initialFormValues, setValue]);
 
+  // Set the selected signal if this is a signal component
+  useEffect(() => {
+    if (!initialFormValues) return;
+    const internalTable =
+      initialFormValues.component?.moped_components?.feature_layer
+        ?.internal_table;
+    const isSignalComponent = internalTable === "feature_signals";
+    if (!isSignalComponent) return;
+    if (!areSignalOptionsLoaded) return;
+
+    const componentSignal = initialFormValues.component?.feature_signals?.[0];
+    const knackFormatSignalOption =
+      featureSignalsRecordToKnackSignalRecord(componentSignal);
+
+    setValue("signal", knackFormatSignalOption);
+  }, [initialFormValues, areSignalOptionsLoaded, setValue]);
+
```

Comment:
this util is working well, but if we look to add many more fields to this form we might want to revisit this to see if we can cut down on the useEffects. i've never understood RFH's preferred way of handling this very common edit situation where your form defaults may not be ready on first render, but i _think_ an alternative is to memoize the the object you pass to RFH's `defaultValues` - and the default form values will update until the form is dirtied. that might offer a way to combine some of the up-front data munging and not have to call `setValue` a whole bunch. it's not important right now; just had to kick in my two cents.

Reply:
@johnclary - glad that you caught this. I almost broke down and refactored this last night. Those `useEffects` that fire on the `clickedComponent` state updates cause a brief blip of the old values in the form before it closes and some re-renders. You are spot on on a better way to do this, and I think the only field that needs some special care would be the signal field that needs to wait on the async call.

Glad that spotted this too! It felt real messy when I dove back in. 🙏

Reply:
cool cool—will make a backlog issue 🚢